src/ast/rewriter/seq_rewriter.cpp

```diff
@@ -2562,6 +2542,233 @@ expr_ref seq_rewriter::mk_derivative_rec(expr* ele, expr* r) {
     return expr_ref(re().mk_derivative(ele, r), m());
 }
 
+/*
+    Combine two if-then-else expressions in BDD form.
+
+    Definition of BDD form:
+        if-then-elses are pushed outwards
+        and sorted by condition ID (cond->get_id()), from largest on
+        the outside to smallest on the inside.
+        Duplicate nested conditions are eliminated.
+
+    Preconditions:
+    - EITHER k is a binary op code on REs (re.union, re.inter, etc.)
+      and cond is nullptr,
+      OR k is if-then-else (OP.ITE) and cond is the condition.
+    - a and b are in BDD form.
+
+    Postcondition: result is in BDD form.
+        if-then-elses are pushed outwards
+        and sorted by condition ID (cond->get_id()), from largest on
+        the outside to smallest on the inside.
+
+    Uses op cache (memoization) to avoid duplicating work for the same
+    pair of pointers.
+*/
+expr_ref seq_rewriter::combine_ites(decl_kind k, expr* a, expr* b, expr* cond) {
+    expr_ref result(m_op_cache.find(k, a, b, cond), m());
+    if (result) {
+        return result;
+    }
+    SASSERT((k == OP_ITE) == (cond != nullptr));
+    expr *acond = nullptr, *a1 = nullptr, *a2 = nullptr,
+         *bcond = nullptr, *b1 = nullptr, *b2 = nullptr;
+    expr_ref result1(m()), result2(m());
+    if (k == OP_ITE) {
```

Comment:
Note, the BDD case for building ites has to consider comparing all three arguments, e.g,, there are cases for "cond" being an ite and having level of condition compared to level of conditions in a, b.
For example, 
https://github.com/Z3Prover/z3/blob/5f9973d8c4b1cbd86dadb58da8425d5e2acfc1a2/src/math/dd/dd_bdd.cpp#L590

Reply:
Interesting, thanks. I did not know that